src/HttpCacheListener.php

```diff
@@ -273,4 +293,68 @@ public function setVary(Headers $headers)
 
         return $this;
     }
+
+    /**
+     * @param Headers $headers
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setEtag(Headers $headers, HttpResponse $response)
+    {
+        if (! empty($this->cacheConfig['etag'])
+            && (! $headers->has('etag')
+                || ! empty($this->cacheConfig['etag']['override']))
+        ) {
+            $etag = new Header\Etag($this->generateEtag($response));
+            $headers->addHeader($etag);
+        }
+
+        return $this;
+    }
+
+    /**
+     * @param HttpRequest $request
+     * @param HttpResponse $response
+     * @return $this
+     */
+    public function setNotModified(HttpRequest $request, HttpResponse $response)
+    {
+        if (!$request->getHeaders()->has('If-Match') || !$response->getHeaders()->has('Etag')) {
+            return $this;
+        }
+
+        $requestEtags = $request->getHeaders()->get('If-Match')->getFieldValue();
+        $requestEtags = !is_array($requestEtags) ? [$requestEtags] : $requestEtags;
+        $responseEtag = $response->getHeaders()->get('Etag')->getFieldValue();
+
+        if (in_array($responseEtag, $requestEtags) || in_array('*', $requestEtags)) {
+            $response->setStatusCode(304);
+            $response->setContent(null);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Generates an Etag based on the response.
+     *
+     * @param HttpResponse $response
+     * @return string Etag
+     */
+    protected function generateEtag(HttpResponse $response)
+    {
+        $generator = new DefaultEtagGenerator();
+
+        // Use custom generator.
+        if (!empty($this->container)
+            && !empty($this->cacheConfig['etag']['generator'])
+            && $this->container->has($this->cacheConfig['etag']['generator'])
+            && $this->container->get($this->cacheConfig['etag']['generator']) instanceof EtagGeneratorInterface
+        ) {
+            $generator = $this->container
+                ->get($this->cacheConfig['etag']['generator']);
+        }
+
+        return $generator->generateEtag($response);
+    }
```

Comment:
To avoid ->get() repetition, usage would be:

```php
// Use custom generator.
if (empty($this->container)
    or empty($this->cacheConfig['etag']['generator'])
    or !$this->container->has($this->cacheConfig['etag']['generator'])
) {
    return;        
}

$generator = $this->container
    ->get($this->cacheConfig['etag']['generator']);

if (!generator instanceof EtagGeneratorInterface) {
    return; // Perhaps throw an exception here?
}    
```